databaker/bake.py

```diff
@@ -197,28 +204,64 @@ def update(self, count):
             sys.stdout.flush()
             self.last_percent = percent
 
-def per_file(fn, recipe, csv):
+def per_file(fn, recipe):
+    filename_params = {"fn": fn.split('.')[0],
+                       "recipe": Opt.recipe_file.split('.')[0]}
     tableset = xypath.loader.table_set(fn, extension='xls')
     showtime("file {!r} imported".format(fn))
+    if Opt.preview:
+        writer = xlutils.copy.copy(tableset.workbook)
+    if Opt.csv:
+        csv_file = Opt.csv_filename.format(**filename_params)
+        csv = TechnicalCSV(csv_file)
     tabs = xypath.loader.get_sheets(tableset, recipe.per_file(tableset))
     for tab_num, tab in enumerate(tabs):
         showtime("tab {!r} imported".format(tab.name))
         obs = recipe.per_tab(tab)
-        obs_count = len(obs)
-        progress = Progress(obs_count, 'Tab {}'.format(tab_num + 1))
-        for ob_num, ob in enumerate(obs):  # TODO use const
-            csv.handle_observation(ob)
-            progress.update(ob_num)
-        print
-
+        if Opt.preview:
+            for i, header in tab.headers.items():
+                if not isinstance(header.bag, xypath.Table):
+                    for bag in header.bag:
+                        writer.get_sheet(tab.index).write(bag.y, bag.x, bag.value,
+                            xlwt.easyxf('pattern: pattern solid, fore-colour {}'.format(colourlist[i])))
+                    for ob in obs:
+                        writer.get_sheet(tab.index).write(ob.y, ob.x, ob.value,
+                            xlwt.easyxf('pattern: pattern solid, fore-colour {}'.format(colourlist[OBS])))
+
+
+        if Opt.csv:
+            obs_count = len(obs)
+            progress = Progress(obs_count, 'Tab {}'.format(tab_num + 1))
+            for ob_num, ob in enumerate(obs):  # TODO use const
+                csv.handle_observation(ob)
```

Comment:
At some point maybe explicitly separate ONS code and give it names that it is ONS specific?